src/logics/aave-v2/service.ts

```diff
@@ -109,50 +88,65 @@ export class Service extends common.Web3Toolkit {
           returnData[i * 2]
         );
 
+        const { aTokenAddress, stableDebtTokenAddress, variableDebtTokenAddress } =
+          this.protocolDataProviderIface.decodeFunctionResult('getReserveTokensAddresses', returnData[i * 2 + 1]);
+
         reserveMap[assetAddress] = {
           isSupplyEnabled: isActive && !isFrozen,
           isBorrowEnabled: isActive && !isFrozen && borrowingEnabled,
         };
 
-        const { aTokenAddress, stableDebtTokenAddress, variableDebtTokenAddress } =
-          this.protocolDataProviderIface.decodeFunctionResult('getReserveTokensAddresses', returnData[i * 2 + 1]);
-
         tokenAddresses.push(assetAddress, aTokenAddress, stableDebtTokenAddress, variableDebtTokenAddress);
       });
 
       const tokens = await this.getTokens(tokenAddresses);
 
       for (let i = 0; i < tokens.length; i += 4) {
-        const assetAddress = tokens[i].address;
+        const asset = tokens[i];
+        const aToken = tokens[i + 1];
+        const stableDebtToken = tokens[i + 2];
+        const variableDebtToken = tokens[i + 3];
+
+        const reserveToken: ReserveTokens = {
+          ...reserveMap[asset.address],
+          asset,
+          aToken,
+          stableDebtToken,
+          variableDebtToken,
+        };
 
-        reserveMap[assetAddress].asset = tokens[i];
-        reserveMap[assetAddress].aToken = tokens[i + 1];
-        reserveMap[assetAddress].stableDebtToken = tokens[i + 2];
-        reserveMap[assetAddress].variableDebtToken = tokens[i + 3];
-      }
+        reserveTokens.push(reserveToken);
 
-      this.reserveTokens = Object.values(reserveMap);
+        reserveMap[asset.address] = reserveToken;
 
-      // Add aToken address as key for quick lookup
-      for (const reserve of Object.values(reserveMap)) {
-        reserveMap[reserve.aToken.address] = reserve;
+        // Add aToken address as key for quick lookup
+        reserveMap[aToken.address] = reserveToken;
       }
 
+      this.reserveTokens = reserveTokens;
       this.reserveMap = reserveMap;
     }
 
     return { reserveTokens: this.reserveTokens!, reserveMap: this.reserveMap! };
```

Comment:
How about we change the condition at the function beginning to `if (!this.reserveTokens || !this.reserveMap)` to avoid using `this.reserveTokens!` & `this.reserveMap!`?

Reply:
thanks for the suggestion, updated